Projects/Server/Items/Container.cs

```diff
@@ -1512,62 +1512,62 @@ private static Item RecurseFindItemByType(Item current, Type[] types, bool recur
 
     public List<T> FindItemsByType<T>(Predicate<T> predicate) where T : Item => FindItemsByType(true, predicate);
 
+    /// <summary>
+    /// Performs a Breadth-First search through all the <see cref="Item"/>s and
+    /// nested <see cref="Container"/>s within this <see cref="Container"/>.
+    /// </summary>
+    /// <typeparam name="T">Type of objects being searched for</typeparam>
+    /// <param name="recurse">Optional: If true, the search will recursively
+    ///     check any nested <see cref="Container"/>s; otherwise, nested
+    ///     <see cref="Container"/>s will not be searched.</param>
+    /// <param name="predicate">Optional: A predicate to check if the <see cref="Item"/>
+    ///     of type <typeparamref name="T"/> is one of the targets of the search.</param>
+    /// <returns>A list of <see cref="Item"/>s of type <typeparamref name="T"/> that matche the optional <paramref name="predicate"/>.</returns>
     public List<T> FindItemsByType<T>(bool recurse = true, Predicate<T> predicate = null) where T : Item
     {
-      List<T> list = new List<T>();
-      RecurseFindItemsByType(this, recurse, list, predicate);
-
-      return list;
-    }
-
-    private static void RecurseFindItemsByType<T>(Item current, bool recurse, List<T> list, Predicate<T> predicate)
-      where T : Item
-    {
-      if (current == null || current.Items.Count == 0)
-        return;
-
-      List<Item> items = current.Items;
-
-      for (int i = 0; i < items.Count; ++i)
+      var queue = new Queue<Container>();

```

Comment:
I don't like the idea of making a queue (or any data structure just for searching. This is a hot path too.
Maybe a pool of queues?

Reply:
Ok so I've created a simple RefPool object that can be reused. It generates Disposable QueueRef objects that will automatically return themselves to the pool when disposed of. Figured it was the safest way to implement. I have it set to retain a max of 10 queues.